bin/get-video-s3-acls.py

```diff
@@ -0,0 +1,325 @@
+            ] + dict_row
+        else:
+            this_all_keys_dict[video_key] = [
+                False,  # NZSL Absent
+                True,  # S3 PRESENT
+            ] + [""] * 6
+
+    # Find NZSL keys that are absent from S3 (present handled above)
+    for video_key, dict_row in this_nzsl_raw_keys_dict.items():
+        if video_key not in this_s3_bucket_raw_keys_list:
+            this_all_keys_dict[video_key] = [
+                True,  # NZSL PRESENT
+                False,  # S3 Absent
+            ] + dict_row
+
+    return this_all_keys_dict
+
+
+# Cases
+# In S3     In NZSL     Action
+#   Is        Not         Delete S3 Object
+#   Is        Is          Update ACL
+#   Not       Is          Review
+#      Other              Review
+def get_recommended_action(key_in_nzsl, key_in_s3):
+    if key_in_s3:
+        if key_in_nzsl:
+            return "Update ACL"
+        else:
+            return "Delete S3 Object"
+    return "Review"
+
+
+# Get S3 object's ACL
+def get_s3_canned_acl(video_key):
+    # TODO pass in a boto client instead of recreating one each time
+    s3_client = boto3.client("s3")
+    acls_grants = s3_client.get_object_acl(Bucket=AWS_S3_BUCKET, Key=video_key)[
+        "Grants"
+    ]
+    if len(acls_grants) > 1:
+        if (
+            acls_grants[0]["Permission"] == "FULL_CONTROL"
+            and acls_grants[1]["Permission"] == "READ"
+        ):
+            return "public-read"
+    elif acls_grants[0]["Permission"] == "FULL_CONTROL":
+        return "private"
+
+    return "unknown"
+
+
+# Get S3 object's LastModified date/time
+def get_s3_lastmodified(video_key):
+    # TODO pass in a boto client instead of recreating one each time
+    return boto3.client("s3").head_object(Bucket=AWS_S3_BUCKET, Key=video_key)[
+        "LastModified"
+    ]
+
+
+def build_csv_header():
+    return CSV_DELIMITER.join(
+        [
+            "Action",
+            "S3 Video key",
+            "S3 LastModified",
+            "S3 Expected Canned ACL",
+            "S3 Actual Canned ACL",
+            "Sbank Gloss ID",
+            "Sbank Video ID",
+            "Sbank Gloss public",
+            "Sbank Video public",
+            "Sbank Gloss",
+            "Sbank Gloss created at",
+        ]
+    )
+
+
+def build_csv_row(
+    video_key,
+    key_in_nzsl=False,
+    key_in_s3=False,
+    gloss_idgloss=None,
+    gloss_created_at=None,
+    gloss_id=None,
+    video_id=None,
+    gloss_public=False,
+    video_public=False,
+):
+    # See signbank/video/models.py, line 59, function set_public_acl()
+    canned_acl_expected = ""
+    if key_in_nzsl:
+        canned_acl_expected = "public-read" if video_public else "private"
+
+    lastmodified = ""
+    canned_acl = ""
+    if key_in_s3:
+        lastmodified = get_s3_lastmodified(video_key)
+        canned_acl = get_s3_canned_acl(video_key)
+
+    action = get_recommended_action(key_in_nzsl, key_in_s3)
+
+    return CSV_DELIMITER.join(
```

Comment:
Issue: Use a csv writer

If and of the fields containers a comma character then your CSV output will not longer be correctly formatted. Instead use the built in csv writer with the output set to `stdout`, this will ensure escaping is done as needed.

```python
out = csv.writer(sys.out)
out.writerow([action, ...])
...

Reply:
I strip out the commas at the input stage, but your observation stands.
I'll see if the docker image has this available, that's one of the constraints on this.

Reply:
Of course it's there, it's built in 😆

Reply:
Done for `get_`, works nicely! I'll do the same for the other two scripts.
We can fiddle with `csv` parameters then later, if we need to.